File: java/save-command/src/main/java/org/babyfish/jimmer/example/save/model/TreeNode.java
package org.babyfish.jimmer.example.save.model;

import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@Entity
@KeyUniqueConstraint
public interface TreeNode {

    @Id
    @Column(name = "node_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @Key
    @Nullable
    @ManyToOne
    @JoinColumn(name = "parent_id")
    @OnDissociate(DissociateAction.DELETE)
    TreeNode parentNode();

    @OneToMany(mappedBy = "parentNode")
    List<TreeNode> childNodes();
}

File: kotlin/save-command-kt/src/main/kotlin/org/babyfish/jimmer/example/save/model/TreeNode.kt
package org.babyfish.jimmer.example.save.model

import org.babyfish.jimmer.sql.*

@Entity
@KeyUniqueConstraint
interface TreeNode {

    @Id
    @Column(name = "node_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    @Key
    val name: String

    @Key
    @ManyToOne
    @JoinColumn(name = "parent_id")
    @OnDissociate(DissociateAction.DELETE)
    val parentNode: TreeNode?

    @OneToMany(mappedBy = "parentNode")
    val childNodes: List<TreeNode>
}